File: JavaScript/923-3Sum With Multiplicity.js
/**
 * @param {number[]} arr
 * @param {number} target
 * @return {number}
 */
var threeSumMulti = function(arr, target) {
    let hmap = new Map();
    let count = 0;
    
    for (let char of arr){
        hmap.set(char, (hmap.get(char)||0) + 1 );
    }
    
    for (let num1 of hmap.keys()){
        for (let num2 of hmap.keys()){
            let num3 = target - num1 - num2;
            if (!hmap.has(num3)){
                continue;
            }
            let count1 = hmap.get(num1), count2 = hmap.get(num2), count3 = hmap.get(num3);
            if (num1 === num2 && num2 === num3){
                count += count1*(count1-1)*(count1-2)/6;
            }else if(num1 === num2){
                count += count3*count1*(count1-1)/2;
            }else if (num1 < num2 && num2 < num3){
                count += count1*count2*count3;
            }
        }
    }
    return count%(Math.pow(10,9) + 7);  
};

File: 923-3Sum With Multiplicity.py
class Solution(object):
    def threeSumMulti(self, arr, target):
        """
        :type arr: List[int]
        :type target: int
        :rtype: int
        """
# Two pointer
#         arr.sort()
#         res = 0
#         for i in range(len(arr)-2):
#             j = i+1
#             k = len(arr)-1
#             while j<k:
#                 if arr[j] + arr[k] < target - arr[i]:
#                     j+=1
#                 elif arr[j] + arr[k] > target - arr[i]:
#                     k-=1
#                 else:
#                     left, right = 1, 1
#                     while j + left < k and arr[j] == arr[j+left]:
#                         left+=1
#                     j = j+left-1
                    
#                     while k-right > j and arr[k] == arr[k-right]:
#                         right+=1
#                     k = k-right+1
                    
#                     if arr[j] == arr[k]:
#                         res += (left + right)*(left+right-1)//2
#                     else:
#                         res += left*right
#                     j+=1
#                     k-=1
        
#         return res % (10**9 + 7)
    
# HashMap

#         hmap = dict()
#         count = 0
        
#         for i in range(1,len(arr)):
#             if target - arr[i] in hmap:
#                 count = (count + hmap[target-arr[i]])%(10**9+7)
#             for j in range(i):
#                 hmap[arr[i]+arr[j]] = hmap.get(arr[i]+arr[j], 0) + 1
        
#         return count
    
# HashMap2

        hmap = dict()
        count = 0
        
        for char in arr:
            hmap[char] = hmap.get(char, 0) + 1
            
        for num1 in hmap.keys():
            for num2 in hmap.keys():
                num3 = target - num1 - num2
                if num3 not in hmap:
                    continue
                if num1 == num2 and num2 == num3:
                    count += hmap[num1]*(hmap[num1]-1)*(hmap[num1]-2)//6
                elif num1 == num2:
                    count += hmap[num3]*hmap[num1]*(hmap[num1]-1)//2
                elif num1 < num2 and num2 < num3:
                    count += hmap[num1]*hmap[num2]*hmap[num3]
        
        return count % (10**9 + 7)